sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/group/ThroughputGroupControllerBase.java

```diff
@@ -0,0 +1,239 @@
+        ThroughputControlGroup group,
+        Integer maxContainerThroughput,
+        RxPartitionKeyRangeCache partitionKeyRangeCache,
+        String targetContainerRid) {
+
+        checkNotNull(globalEndpointManager, "Global endpoint manager can not be null");
+        checkNotNull(group, "Throughput control group can not be null");
+        checkNotNull(partitionKeyRangeCache, "Partition key range cache can not be null or empty");
+        checkArgument(StringUtils.isNotEmpty(targetContainerRid), "Target container rid cannot be null nor empty");
+
+        this.connectionMode = connectionMode;
+        this.globalEndpointManager = globalEndpointManager;
+        this.group = group;
+
+        if (this.group.getTargetThroughputThreshold() != null) {
+            checkNotNull(maxContainerThroughput, "Max container throughput can not be null when target throughput threshold defined");
+            this.maxContainerThroughput = new AtomicInteger(maxContainerThroughput);
+        } else {
+            this.maxContainerThroughput = null;
+        }
+
+        this.groupThroughput = new AtomicReference<>();
+        this.calculateGroupThroughput();// Calculate after maxContainerThroughput
+
+        this.partitionKeyRangeCache = partitionKeyRangeCache;
+        this.requestControllerAsyncCache = new AsyncCache<>();
+        this.targetContainerRid = targetContainerRid;
+
+        this.cancellationTokenSource = new CancellationTokenSource();
+    }
+
+    private void calculateGroupThroughput() {
+        double allocatedThroughput = Double.MAX_VALUE;
+        if (this.group.getTargetThroughputThreshold() != null) {
+            allocatedThroughput = Math.min(allocatedThroughput, this.maxContainerThroughput.get() * this.group.getTargetThroughputThreshold());
+        }
+
+        if (this.group.getTargetThroughput() != null) {
+            allocatedThroughput = Math.min(allocatedThroughput, this.group.getTargetThroughput());
+        }
+
+        this.groupThroughput.set(allocatedThroughput);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> Mono<T> init() {
+        return this.resolveRequestController()
+            .doOnSuccess(dummy -> {
+                this.throughputUsageCycleRenewTask(this.cancellationTokenSource.getToken()).subscribeOn(Schedulers.parallel()).subscribe();
+            })
+            .thenReturn((T)this);
+    }
+
+    private Flux<Void> throughputUsageCycleRenewTask(CancellationToken cancellationToken) {
+        checkNotNull(cancellationToken, "Cancellation token can not be null");
+        return Mono.delay(DEFAULT_THROUGHPUT_USAGE_RESET_DURATION)
+            .flatMap(t -> this.resolveRequestController())
+            .doOnSuccess(requestController -> requestController.renewThroughputUsageCycle(this.groupThroughput.get()))
+            .onErrorResume(throwable -> {
+                logger.warn("Reset throughput usage failed with reason %s", throwable);
```

Comment:
```suggestion
                logger.warn("Reset throughput usage failed", throwable);
```

in general you should use slf4j place holder "{}" instead of "%s". for exceptions slfj4 has special support if they come as the last arg. no need for any place holder

Reply:
got it, will change across everywhere.